### PR TITLE
Integrate CRUD actions with question service

### LINKED ISSUE #32: [Question Service] Integration issues and clarifications
Below are some complications faced during integration:

### Issues: 
- [x] mongodb not in backend package.json? module not found when starting service.
- [ ] allow more char for question description.
with the addition of a rich text editor, adding images can easily exceed the current word limit. Need help to relax the limit.
image
image

### Clarifications: 
- [ ] question id or _id? 
the assignment seems to follow the leetcode style of using natural numbers for question id, the backend is currently using the MongoDB way of keeping ids. So which style should we follow moving ahead?
image  image
- [x] Structure of response format
  The question data is currently nested in a separate question field:
  ```
  {
      _id:
      question: {
           ...
      }
  }
  ``` 
  Can the backend return flattened format? else the frontend have to implement a separate type to match the structure.
  ```
  {
      id:
      title:
      ...
  }
  ``` 
- [x] Separation of examples and constraints from the description.
Do we want to split the fields? It's manageable for now, but we might face issues if we are going to implement assignment 6 (leetcode crawler)

### LINKED ISSUE #36: Integrate question service to perform CRUD for questions

### PR DESCRIPTION
**Pull request by c0j0s**
# Pull Request

## Description

Add CRUD support for question service.

## Related Issue(s)

Close #32 
Close #36 

## Changes Made

- Redo create form to support constrain and example values.
- Hide id field for now, pending removal if necessary.
- Update question table layout and actions.
- Add rich text editor support for description field.
- Changes to endpoint.ts, `api()` now returns `ApiResponse` type for wrapper layer to access.

### Retrieve and Delete
image

### Create and edit
image

## Checklist
- [x] I have checked that the changes included in the PR are intended to merge to `master` or any destination branch.
- [x] I have verified that the new changes do not break any existing functionalities, unless the new changes are intended and have approved by the team.
- [x] I will take care of the merging and delete the side-branch after the PR is merged.

## Additional Notes/References

- Description field needs to relax the word limit, see #32.
- Rich text editor: see [this](https://www.npmjs.com/package/react-quill#with-webpack-or-create-react-app)

### CODE REVIEW

**Review by weidak — APPROVED** [review]
Approving this so that I can utilize the updated endpoint.ts and various enums js has created for my user integration

**Comment by weidak** [comment]
I think the rich text editor is a good addition and yes lets just keep away the ID field for now since we can always change it later on.

image

Is it possible that we add an (optional) explanation field here so that we can explain the examples? I think thats already inside the schema of the db too. 

Leetcode:
image

If we do add the explanation part, we can consider moving the examples/explanations to below the constraints (or vice versa) so that we have more space to work with (col instead of row) cos it looks like it will be cramped. 

Overall it looks great 👍🏻 thanks junsheng for the hard work. I think for the MVP this is really good.

File: frontend/src/helpers/endpoint.ts [hunk 1]
```diff
@@ -1,27 +1,43 @@
...
+ */
+export default async function api(config: ApiConfig): Promise<ApiResponse> {
+  // Configure gateway host based on the environment (production or development).
   const host =
     process.env.NODE_ENV == "production"
       ? process.env.ENDPOINT_PROD
       : process.env.ENDPOINT_DEV;
 
+  // Configure local service port based on the 'service' property in the configuration.
   let servicePort = ":";
```
Comment (tryyang2001) [inline frontend/src/helpers/endpoint.ts:41]: I have created an issue [here ](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/issues/41)asking for the same question: I was curious if we need to always assign a port number before we call an endpoint. I thought we could pre-defined it and then keep using it in each service?

File: frontend/src/helpers/endpoint.ts [hunk 2]
```diff
@@ -51,15 +72,25 @@ export const api = async (config: apiConfig) => {
...
-      return [];
-    }
+    // Parse the response body for all status codes except 204 (no content).
+    let data = res.status != 204 ? await res.json() : {};
+
+    logger.info(`[${res.status}] ${config.method}: ${res.url}}`);
 
-    return res.json();
+    // Return an ApiResponse object with status, data, and message.
+    return {
```
Comment (tryyang2001) [inline frontend/src/helpers/endpoint.ts:81]: Was curious why not we just return the response directly? Then we also have access to status code, status text, data, as well as other things like the response headers and etc. if we need them.

File: frontend/src/types/question.d.ts [hunk 3]
```diff
@@ -1,9 +1,16 @@
...
+  output: string;
+};
+
 type Question = {
-  id: string;
+  _id?: string;
   title: string;
   complexity: string;
-  topics: Set<string>;
+  topics: string[];
```
Comment (tryyang2001) [inline frontend/src/types/question.d.ts:10]: Oh, why don't you just declare it as `Topic[]`?

**Comment by tryyang2001** [comment]
Sorry that I created and merged this PR which updates the enum in frontend. I noticed the `Service` enum string value is not supposed to be all uppercase, can you change it back (as well as the user service) when you merge this PR? Thank you!

File: frontend/src/helpers/endpoint.ts (hunk 2 above)
Comment (c0j0s) [reply frontend/src/helpers/endpoint.ts:81]: here it returns an ApiResponse object which includes some of the mentioned values. the intended purpose is to simplify the handling after returning to the wrapper layer. for instance the response body, some request does not return any payload, so calling response.json() will throw an exception which is not that friendly for warpper functions to handle. returning ApiResponse hides some of the complexity within the endpoint layer. so this is the idea, but we can change it to return the response object directly if this approach is not working well with other services.

File: frontend/src/types/question.d.ts (hunk 3 above)
Comment (c0j0s) [reply frontend/src/types/question.d.ts:10]: hmm... don't see the need as topics are just an arr of string.

File: frontend/src/helpers/endpoint.ts (hunk 2 above)
Comment (tryyang2001) [reply frontend/src/helpers/endpoint.ts:81]: Ok got it, as I don't see any potential issue for now. I am fine with your approach, so I think there is no need to change it.

File: frontend/src/types/question.d.ts (hunk 3 above)
Comment (tryyang2001) [reply frontend/src/types/question.d.ts:10]: Oh, but I thought inside this topics array, we only accept `Topic` (which is an enum) value?

**Comment by c0j0s** [comment]
> Is it possible that we add an (optional) explanation field here so that we can explain the examples? I think thats already inside the schema of the db too.
> If we do add the explanation part, we can consider moving the examples/explanations to below the constraints (or vice versa) so that we have more space to work with (col instead of row) cos it looks like it will be cramped.
> 

image  


I'm not sure if there's a standard format for examples in leetcode so the purpose of inputting examples in their own table here is more like creating test cases for the collab service if we decide to implement online judging later on.  As for explanations, i don't see a purpose of handling them separately, that is why I included a rich text editor, so pictures and other non-structure information can be placed together.

**Comment by weidak** [comment]
> > Is it possible that we add an (optional) explanation field here so that we can explain the examples? I think thats already inside the schema of the db too.
> 
> > If we do add the explanation part, we can consider moving the examples/explanations to below the constraints (or vice versa) so that we have more space to work with (col instead of row) cos it looks like it will be cramped.
> 
> > 
> 
> 
> 
> image  
> 
> 
> 
> 
> 
> I'm not sure if there's a standard format for examples in leetcode so the purpose of inputting examples in their own table here is more like creating test cases for the collab service if we decide to implement online judging later on.  As for explanations, i don't see a purpose of handling them separately, that is why I included a rich text editor, so pictures and other non-structure information can be placed together. 

Ok sure! Didnt realise your train of thought, Im fine with this implementation too

File: frontend/src/types/question.d.ts (hunk 3 above)
Comment (c0j0s) [reply frontend/src/types/question.d.ts:10]: Oh right, but anyway the frontend will be fetching topics from the backend ya? so I guess I won't have the enum in frontend by then.

**Comment by c0j0s** [comment]
> Sorry that I created and merged this [PR ](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/pull/42)which updates the enum in frontend. I noticed the `Service` enum string value is not supposed to be all uppercase, can you change it back (as well as the user service) when you merge this PR? Thank you!

done

File: frontend/src/types/question.d.ts (hunk 3 above)
Comment (tryyang2001) [reply frontend/src/types/question.d.ts:10]: Yes, I think we can use the assumption that the data retrieved from the database should be correct. So we can assume backend will pass valid topics to frontend, hopefully. Then by this assumption, there is no need to put `Topic[]` bah. Let's resolve this conversation :)

**Review by tryyang2001 — APPROVED** [review]
Thanks for updating some common code as well. I will let you decide if you want to update your `api()` method to allow caller to give the request headers, either in this PR or another PR. But I will pre-approve it to allow merging. Meanwhile I think @tlyi is reviewing, you might want to postpone merging for a while. LGTM in general.

**Comment by tlyi** [comment]
Thanks for making the function of the endpoint and wrapper files so clear! Looks good to me, in fact I have learnt a lot from your implementation. Thanks @c0j0s for your hard work !! 👍 😄

**Pull request merged**